// ioc4ts/src/index.ts
/**
 * ioc4ts
 * 
 * 使用 typescript 实现的 IoC 容器
 * 
 * @author zhangbaili
 * @since 2021/10/23
*/

import Get from './web/annotations/Get'
import ReturnType from './core/annotations/ReturnType'
import Property from './core/annotations/Property'
import PathVariable from './web/annotations/PathVariable'
import Post from './web/annotations/Post'
import Put from './web/annotations/Put'
import Delete from './web/annotations/Delete'
import RequestParam from './web/annotations/RequestParam'
import RequestHeader from './web/annotations/RequestHeader'
import RequestBody from './web/annotations/RequestBody'
import ApplicationContext from './core/ApplicationContext'
import HttpRequest from './web/annotations/HttpRequest'
import HttpRequestInterface from './web/HttpRequestInterface'
import View from './view/annotations/View'
import WebApplication from './view/WebApplication'

export {
    Get,
    ReturnType,
    Property,
    PathVariable,
    Post,
    Put,
    Delete,
    RequestParam,
    RequestHeader,
    RequestBody,
    HttpRequest,
    HttpRequestInterface,
    View,
    WebApplication
}

export default ApplicationContext
