// ioc4ts/build/src/core/ApplicationContext.js
"use strict";
/**
 * ApplicationContext 类
 *
 * 应用程序的上下文
 *
 * @author zhangbaili
 * @since 2021/10/24 19:26
*/
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const BeanFactory_1 = __importDefault(require("./BeanFactory"));
const ViewBeanRegistry_1 = __importDefault(require("./ViewBeanRegistry"));
class ApplicationContext {
    constructor(config) {
        var _a;
        this.log = (_a = config === null || config === void 0 ? void 0 : config.log) !== null && _a !== void 0 ? _a : false;
        this.beanFactory = new BeanFactory_1.default();
        this.viewBeanRegistry = new ViewBeanRegistry_1.default();
    }
    static getInstance(config) {
        if (!this.instance) {
            this.instance = new ApplicationContext(config);
        }
        return this.instance;
    }
    getBeanFactory() {
        return this.beanFactory;
    }
    getViewBeanRegisty() {
        return this.viewBeanRegistry;
    }
    isLog() {
        return this.log;
    }
}
exports.default = ApplicationContext;
