// ioc4ts/src/core/ApplicationContext.ts
/**
 * ApplicationContext 类
 * 
 * 应用程序的上下文
 * 
 * @author zhangbaili
 * @since 2021/10/24 19:26
*/

import BeanFactory from './BeanFactory'
import ViewBeanRegistry from './ViewBeanRegistry'

interface ApplicationContextConfig {
    log: boolean
}

export default class ApplicationContext {
    private static instance: ApplicationContext
    private log: boolean
    private beanFactory: BeanFactory
    private viewBeanRegistry: ViewBeanRegistry
    
    private constructor(config?: ApplicationContextConfig) {
        this.log = config?.log ?? false
        this.beanFactory = new BeanFactory()
        this.viewBeanRegistry = new ViewBeanRegistry()
    }

    static getInstance(config?: ApplicationContextConfig) {
        if(!this.instance) {
            this.instance = new ApplicationContext(config)
        }

        return this.instance
    }

    getBeanFactory() : BeanFactory {
        return this.beanFactory
    }

    getViewBeanRegisty() {
        return this.viewBeanRegistry
    }

    isLog () {
        return this.log
    }

}
