// ioc4ts/build/src/index.js
"use strict";
/**
 * ioc4ts
 *
 * 使用 typescript 实现的 IoC 容器
 *
 * @author zhangbaili
 * @since 2021/10/23
*/
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.View = exports.HttpRequest = exports.RequestBody = exports.RequestHeader = exports.RequestParam = exports.Delete = exports.Put = exports.Post = exports.PathVariable = exports.Property = exports.ReturnType = exports.Get = void 0;
const Get_1 = __importDefault(require("./web/annotations/Get"));
exports.Get = Get_1.default;
const ReturnType_1 = __importDefault(require("./core/annotations/ReturnType"));
exports.ReturnType = ReturnType_1.default;
const Property_1 = __importDefault(require("./core/annotations/Property"));
exports.Property = Property_1.default;
const PathVariable_1 = __importDefault(require("./web/annotations/PathVariable"));
exports.PathVariable = PathVariable_1.default;
const Post_1 = __importDefault(require("./web/annotations/Post"));
exports.Post = Post_1.default;
const Put_1 = __importDefault(require("./web/annotations/Put"));
exports.Put = Put_1.default;
const Delete_1 = __importDefault(require("./web/annotations/Delete"));
exports.Delete = Delete_1.default;
const RequestParam_1 = __importDefault(require("./web/annotations/RequestParam"));
exports.RequestParam = RequestParam_1.default;
const RequestHeader_1 = __importDefault(require("./web/annotations/RequestHeader"));
exports.RequestHeader = RequestHeader_1.default;
const RequestBody_1 = __importDefault(require("./web/annotations/RequestBody"));
exports.RequestBody = RequestBody_1.default;
const ApplicationContext_1 = __importDefault(require("./core/ApplicationContext"));
const HttpRequest_1 = __importDefault(require("./web/annotations/HttpRequest"));
exports.HttpRequest = HttpRequest_1.default;
const View_1 = __importDefault(require("./view/annotations/View"));
exports.View = View_1.default;
exports.default = ApplicationContext_1.default;
